contrib/zenith/file_cache.c

```diff
@@ -0,0 +1,345 @@
+
+	if (prev_shmem_startup_hook)
+	{
+		prev_shmem_startup_hook();
+	}
+
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+
+	lfc_ctl = (FileCacheControl*)ShmemInitStruct("lfc", sizeof(FileCacheControl), &found);
+	if (!found)
+	{
+		lfc_lock = (LWLockId)GetNamedLWLockTranche("lfc_lock");
+		info.keysize = sizeof(BufferTag);
+		info.entrysize = sizeof(FileCacheEntry);
+		lfc_hash = ShmemInitHash("lfc_hash",
+								 /* lfc_size+1 because we add new element to hash table before eviction of victim */
+								 lfc_size+1, lfc_size+1,
+								 &info,
+								 HASH_ELEM | HASH_BLOBS);
+		lfc_ctl->size = 0;
+		lfc_prune(&lfc_ctl->lru);
+
+		/* Remove file cache on restart */
+		(void)unlink(lfc_path);
+	}
+	LWLockRelease(AddinShmemInitLock);
+}
+
+void
+lfc_init(void)
+{
+	/*
+	 * In order to create our shared memory area, we have to be loaded via
+	 * shared_preload_libraries.
+	 */
+	if (!process_shared_preload_libraries_in_progress)
+		elog(ERROR, "Zenith module should be loaded via shared_preload_libraries");
+
+	DefineCustomIntVariable("zenith.file_cache_size",
+							"Maximal size of zenith local file cache (chunks)",
+							NULL,
+							&lfc_size,
+							0, /* disabled by default */
+							0,
+							INT_MAX,
+							PGC_POSTMASTER,
+							0,
+							NULL,
+							NULL,
+							NULL);
+
+	DefineCustomStringVariable("zenith.file_cache_path",
+							   "Path to local file cache (can be raw device)",
+							   NULL,
+							   &lfc_path,
+							   "file.cache",
+							   PGC_POSTMASTER,
+							   0,
+							   NULL,
+							   NULL,
+							   NULL);
+
+	if (lfc_size == 0)
+		return;
+
+	RequestAddinShmemSpace(sizeof(FileCacheControl) + hash_estimate_size(lfc_size+1, sizeof(FileCacheEntry)));
+	RequestNamedLWLockTranche("lfc_lock", 1);
+
+	prev_shmem_startup_hook = shmem_startup_hook;
+	shmem_startup_hook = lfc_shmem_startup;
+}
+
+/*
+ * Try to read page from local cache.
+ * Returns true if page is found in local cache.
+ * In case of error lfc_size is set to zero to disable any further opera-tins with cache.
+ */
+bool
+lfc_read(SMgrRelation reln, ForkNumber forkNum, BlockNumber blkno,
+		 char *buffer)
+{
+	BufferTag tag;
+	FileCacheEntry* entry;
+	ssize_t rc;
+	int chunk_offs = blkno & (CHUNK_SIZE-1);
+
+	if (lfc_size == 0) /* fast exit if file cache is disabled */
+		return false;
+
+	tag.rnode = reln->smgr_rnode.node;
+	tag.forkNum = forkNum;
+	tag.blockNum = blkno & ~(CHUNK_SIZE-1);
+
+	LWLockAcquire(lfc_lock, LW_EXCLUSIVE);
+	entry = hash_search(lfc_hash, &tag, HASH_FIND, NULL);
+	if (entry == NULL || (entry->bitmap[chunk_offs >> 5] & (1 << (chunk_offs & 31))) == 0)
+	{
+		/* Page is not cached */
+		LWLockRelease(lfc_lock);
+		return false;
+	}
+	/* Unlink entry from LRU list to pin it for the duration of IO operation */
+	if (entry->access_count++ == 0)
+		lfc_unlink(entry);
+	LWLockRelease(lfc_lock);
+
+	/* Open cache file if not done yet */
+	if (lfc_desc == 0)
+	{
+		lfc_desc = BasicOpenFile(lfc_path, O_RDWR|O_CREAT);
```

Comment:
While you're trading memory for file system; why do you still allow the OS to buffer these files by not using O_DIRECT? This leads to double buffering, which limits the effectiveness of the file system as a replacement for shared memory buffers.

Reply:
It is long story with double buffering... Some people think that is awful thing and should be avoided as much as possible. I do not completely agree. 
I have made many experiments in the past (may be situation changed now) and results with file system cache was always better then without it (relying only on user mode cache).
Only OS knows how much free memory it actually has and so how much of it can be used for file system cache. Any attempt to guess it at application level are not precise. And either you are not using all resources and di extra IO, either (much worse) when cache is swapped to the disk and instead of speed-up cause only extra trashing.
So if you do not know for sure memory footprint of all your application, it is not good idea to play with DIRECT_IO.
Especially in cloud, with multiple level of virtualization, it is better not to play with own memory management. In this case you should specify small enough shared_buffers and hope that file system will do the best to cache data. Fortunately Vanilla Postgres performance with small shared buffers is not significantly worser than with larger file buffers.

It is not true in case of Zenith, because get_page roundtrip is much more expensive than read from OS cache. This is why file cache can be very useful. And we definitely do not need to prevent OS from making it even more efficient because of caching. And if shared_buffers are small (few Mb), then there will be really no double buffering, because we "double" only this few Mb.

Reply:
> case you should specify small enough shared_buffers and hope that file system will do the best to cache data

But then you're dependent on the file system cache to actually do it's job, which cannot be guaranteed, especially when running in a shared environment such as kubernetes (which usually shares a file system buffer pool across pods). One bad sibling will trash performance for other nodes; and I don't like that: I'd take consistent performance over inconsistent performance every day.

I don't really get why we would want small (few megabytes) shared buffers. That seems like just enough to start PG, but not enough to run anything meaningful. A clean database with dbinit is already 10s of megabytes, and even RDS's smallest PG nodes have at least 1GB of memory / 250MiB shared buffers.

Reply:
(other point of concern: this effectively makes the 'main' page cache LRU instead of PGs own eviction algorithm. I think PG had good reasons to move from LRU to the current algoritm, I'm not sure that simple LRU is the right option. Yes, the eviction algorithm should be relatively easily changed, but I hope my point is clear)

Reply:
> I'd take consistent performance over inconsistent performance every day.
You will never get consistent performance especially in shared environment. Even at your personal desktop, with no other programs active, speed of your program may vary several times depending on how data is located in memory. Certainly I am talking about some special cases. But it is real naive to expect that you will get consistent performance if you will use direct io. Quite opposite, attempts to take control over system resources (bind tasks to CPUs, pin memory regions, use direct IO) may lead to bad and unpredicable results.

Reply:
The main advantage of Clock comparing with LRU is better scalability. But LRU usually provide better hit rate. I am not insisting that LRU is the perfect choice for local file cache. I just want to notice that in my tests I do not see much lock contention. So I prefer not to solve unexisted problems. Definitely it doesn't mean that we will not see such contention on some workloads. I will be pleased if you can propose some test demonstrating such contention. 

Please notice that the difference between postgres shared buffers and local files system cache is that in case of page hit in first case we just get access to the page (and may be just read single word from this page), and in case of file cache - read page from the disk. Obviously, the difference in times can be several orders of magnitude.

Reply:
The cache replacement algorithm probably makes very little difference in this case. I mean that it is not very important that you replace the best possible page. This cache sits behind the Postgres buffer cache, so none of the pages in it are very hot anyway. In fact, I'd suggest using random policy, it's simple and scales very well.

> I just want to notice that in my tests I do not see much lock contention. So I prefer not to solve unexisted problems.

+1

Reply:
> But then you're dependent on the file system cache to actually do it's job, which cannot be guaranteed, especially when running in a shared environment such as kubernetes (which usually shares a file system bufnfer pool across pods). One bad sibling will trash performance for other nodes; and I don't like that: I'd take consistent performance over inconsistent performance every day.

On the other hand, if you share the cache among the tenants, then you can dynamically use the memory where it's needed the most. In a nutshell, if you have one instance that is idle, or has a workload that uses very little, we can make that memory available to other tenants that have more use for it. Yeah, you are then exposed to noisy neighbours, and performance can be inconsistent. It's a tradeoff, I think we'll need to experiment with different settings to get the balance right.

> I don't really get why we would want small (few megabytes) shared buffers. That seems like just enough to start PG, but not enough to run anything meaningful. A clean database with dbinit is already 10s of megabytes, and even RDS's smallest PG nodes have at least 1GB of memory / 250MiB shared buffers.

A few megabytes is probably too small, the main reason we do that currently is to expose bugs more quickly. But I think a 32 MB or 64 MB shared buffers might be feasible. Depends on how fast we can make this filesystem cache.

Reply:
I suspect nether plain LRU nor LFU are good enough algorithms as they ignore one the properties of data, there's no preference to the page usefulness, e.g. index root/meta pages are as likely to be evicted as index leaf pages (assuming equal timestamp in case of LRU, or equal use count in case of LFU).

I suggest we should implement a Score function that takes more than just buffer specific information (e.g. N last access times, use count, page type) but also relation type, number of cached index/table pages

Reply:
> On the other hand, if you share the cache among the tenants, then you can dynamically use the memory where it's needed the most. In a nutshell, if you have one instance that is idle, or has a workload that uses very little, we can make that memory available to other tenants that have more use for it. Yeah, you are then exposed to noisy neighbours, and performance can be inconsistent. It's a tradeoff, I think we'll need to experiment with different settings to get the balance right.

+1 on the multi-tenant cache, and I think it may make sense to have another such cache in the RAM, it would allow implementation of prefetch and pretty nice performance even on small instances.

Reply:
1. Postgres eviction policy is actually very far from ideal. Page access counter is limited by 5. So if some heap page is updated 5 times, it get the highest possible rate and compete with much more frequently updated pages for eviction. As far as heap size is usually much larger than index size, it cause flushing index pages from the cache. So I think that smart eviction algorithm underneath Postgres shared buffers still make sense and will be better than random eviction.
2. Shared in-memory cache will require communication by  sockets between VM. How else can shared cache can be implemented for Postgres compute nodes executed each in its own VM? I do not think that such cache will be much better than pageserver. At least right now with local Zenith installation (when pageserver and compute nodes are on the same host) we still suffer from large get_page latency.
3. Neither LRU, LFU... any other eviction algorithm is perfect. And for each it is possible to find bad access pattern. There were a lot of different PHD works on this topic. Some time ago I have initiated discussion about improving Postgres cache replacement algorithm: https://www.postgresql.org/message-id/flat/CA%2BTgmobLDVTsCfGOOA8HN1s5LjQCGfOCLK7LQgB-vXfJmaQJZg%40mail.gmail.com#f7a5ef324e98e4210cd8f48aa274800d  I even have implemented some prototype but the difference in performance with vanilla postgres was negligible. I doesn't mean that choice of eviction algorithm is not important. I just want to say that difference of any proposed ***+++ algrorithm  and "classical" algorithm like LRU,clock ... is not so large. And for each of them it is possible to find workload on which competitors will be better. I do no think that we should spent time trying to invent our on bike<---algirthm, unless somebody whats to defeat PHD:)

Right now, current "naive" implementation of local file system cache shows performance comparable with shared buffers when dataset fits in host memory. So  I wonder what else you want to improve. Try to make it faster than shared buffers?

Reply:
> Yeah, you are then exposed to noisy neighbours, and performance can be inconsistent. It's a tradeoff, I think we'll need to experiment with different settings to get the balance right.

Do we want to optimize for cost first, or for performance first?
Do we support customers tweaking their database settings for performance, or is that considered bad customer behaviour?

I can't see us competing with Aurora on anything other than 'open source code base' if the only option we give customers is "Don't have a database larger than a few hundred megabytes, or your performance cannot be guaranteed because we promote noisy neighbours in your neighbourhood".

It's not difficult to guarantee CPU and memory for each compute node. Yes, that is wasted resources if the compute node is not in active use, but it's not "wasted" if that provides much more consistent performance, or (at least) better worst cases when noisy neighbours are considered, or won't randomly OOM compute that are using the available memory.